File: anansi-frontend-fix/src/stores/integrations.ts
/**
 * Integrations Store — Zustand state management for connector connections.
 *
 * Manages:
 * - List of available connectors with status
 * - Active integration details
 * - Webhook registrations
 * - Connection state (connecting, connected, error)
 */

import { create } from "zustand";
import { api } from "../lib/api";
import type { Integration, Connector } from "../types";

// ─── Types ─────────────────────────────────────────────────────────────────

export type ConnectionStatus = "connected" | "disconnected" | "error" | "pending";

export interface ConnectorWithStatus extends Connector {
  status: ConnectionStatus;
  integrationId?: string;
  lastSyncAt?: string | null;
  errorMessage?: string | null;
  scopes?: string[];
}

export interface IntegrationStatusSummary {
  total: number;
  connected: number;
  error: number;
  available: number;
}

interface ConnectionsStatus {
  summary: IntegrationStatusSummary;
  connections: Array<{
    integrationId: string;
    connectorType: string;
    name: string;
    iconUrl: string;
    category: string;
    status: string;
    errorMessage: string | null;
    lastSyncAt: string | null;
  }>;
}

interface WebhookRegistration {
  webhookId: string;
  webhookUrl: string;
  secret: string;
  description: string;
  events: string[];
  active: boolean;
}

interface IntegrationState {
  // ── State ──
  connectors: ConnectorWithStatus[];
  selectedIntegration: Integration | null;
  statusSummary: IntegrationStatusSummary | null;
  isConnecting: boolean;
  isConnected: boolean;
  error: string | null;
  webhooks: WebhookRegistration[];

  // ── Actions ──
  fetchConnectors: () => Promise<void>;
  fetchIntegrationDetail: (id: string) => Promise<Integration | null>;
  initiateOAuth: (connectorKey: string, redirectUri: string) => Promise<{ authUrl: string; state: string } | null>;
  handleOAuthCallback: (connectorKey: string, code: string, state: string) => Promise<boolean>;
  connectApiKey: (connectorKey: string, apiKey: string, config?: Record<string, unknown>) => Promise<boolean>;
  disconnect: (integrationId: string) => Promise<boolean>;
  testConnection: (integrationId: string) => Promise<{ success: boolean; error?: string }>;
  fetchStatus: () => Promise<void>;
  refreshToken: (integrationId: string) => Promise<boolean>;
  registerWebhook: (agentId: string, config: Record<string, unknown>) => Promise<WebhookRegistration | null>;
  unregisterWebhook: (webhookId: string) => Promise<boolean>;
  fetchWebhooks: (agentId?: string) => Promise<void>;
  reset: () => void;
}

// ─── Store ─────────────────────────────────────────────────────────────────

const API_PREFIX = "/api/v1/integrations";

export const useIntegrationStore = create<IntegrationState>((set, get) => ({
  // ── Initial State ──
  connectors: [],
  selectedIntegration: null,
  statusSummary: null,
  isConnecting: false,
  isConnected: false,
  error: null,
  webhooks: [],

  // ── Fetch all connectors with status ──
  fetchConnectors: async () => {
    try {
      set({ error: null, isConnecting: true });
      const data = await api.get<
        Array<{
          key: string;
          name: string;
          description: string | null;
          icon_url: string | null;
          category: string;
          auth_type: string;
          is_connected: boolean;
          status: string;
          integration_id: string | null;
          last_sync_at: string | null;
          error_message: string | null;
        }>
      >(API_PREFIX);

      const connectors: ConnectorWithStatus[] = data.map((c) => ({
        id: c.key,
        key: c.key,
        name: c.name,
        description: c.description ?? null,
        iconUrl: c.icon_url ?? null,
        authType: c.auth_type as "oauth2" | "apikey" | "basic",
        isBuiltin: true,
        status: c.is_connected ? (c.status === "error" ? "error" : "connected") : "disconnected",
        integrationId: c.integration_id ?? undefined,
        lastSyncAt: c.last_sync_at,
        errorMessage: c.error_message,
      }));

      set({ connectors, isConnecting: false });
    } catch (err) {
      set({
        error: err instanceof Error ? err.message : "Failed to fetch connectors",
        isConnecting: false,
      });
    }
  },

  // ── Fetch single integration detail ──
  fetchIntegrationDetail: async (id: string) => {
    try {
      const data = await api.get<Integration>(`${API_PREFIX}/${id}`);
      set({ selectedIntegration: data });
      return data;
    } catch (err) {
      set({ error: err instanceof Error ? err.message : "Failed to fetch integration" });
      return null;
    }
  },

  // ── Initiate OAuth flow ──
  initiateOAuth: async (connectorKey: string, redirectUri: string) => {
    try {
      set({ error: null, isConnecting: true });
      const data = await api.post<{ auth_url: string; state: string; integration_id: string }>(
        `${API_PREFIX}/auth`,
        { connector_key: connectorKey, redirect_uri: redirectUri }
      );
      set({ isConnecting: false });
      return { authUrl: data.auth_url, state: data.state };
    } catch (err) {
      set({
        error: err instanceof Error ? err.message : "Failed to initiate OAuth",
        isConnecting: false,
      });
      return null;
    }
  },

  // ── Handle OAuth callback ──
  handleOAuthCallback: async (connectorKey: string, code: string, state: string) => {
    try {
      set({ error: null, isConnecting: true });
      await api.post(`${API_PREFIX}/auth/callback`, {
        connector_key: connectorKey,
        code,
        state,
      });
      set({ isConnecting: false, isConnected: true });
      await get().fetchConnectors();
      return true;
    } catch (err) {
      set({
        error: err instanceof Error ? err.message : "OAuth callback failed",
        isConnecting: false,
      });
      return false;
    }
  },

  // ── Connect via API key ──
  connectApiKey: async (connectorKey: string, apiKey: string, config?: Record<string, unknown>) => {
    try {
      set({ error: null, isConnecting: true });
      await api.post(`${API_PREFIX}/connect`, {
        connector_key: connectorKey,
        api_key: apiKey,
        config: config ?? {},
      });
      set({ isConnecting: false, isConnected: true });
      await get().fetchConnectors();
      return true;
    } catch (err) {
      set({
        error: err instanceof Error ? err.message : "Failed to connect",
        isConnecting: false,
      });
      return false;
    }
  },

  // ── Disconnect ──
  disconnect: async (integrationId: string) => {
    try {
      set({ error: null });
      await api.post(`${API_PREFIX}/${integrationId}/disconnect`);
      set({ isConnected: false, selectedIntegration: null });
      await get().fetchConnectors();
      return true;
    } catch (err) {
      set({ error: err instanceof Error ? err.message : "Failed to disconnect" });
      return false;
    }
  },

  // ── Test connection ──
  testConnection: async (integrationId: string) => {
    try {
      set({ error: null });
      const result = await api.post<{ success: boolean; error?: string }>(
        `${API_PREFIX}/${integrationId}/test`
      );
      await get().fetchConnectors();
      return result;
    } catch (err) {
      const msg = err instanceof Error ? err.message : "Connection test failed";
      set({ error: msg });
      return { success: false, error: msg };
    }
  },

  // ── Fetch status ──
  fetchStatus: async () => {
    try {
      const data = await api.get<ConnectionsStatus>(`${API_PREFIX}/status`);
      set({ statusSummary: data.summary });
    } catch (err) {
      set({ error: err instanceof Error ? err.message : "Failed to fetch status" });
    }
  },

  // ── Refresh token ──
  refreshToken: async (integrationId: string) => {
    try {
      set({ error: null });
      await api.post(`${API_PREFIX}/${integrationId}/refresh`);
      return true;
    } catch (err) {
      set({ error: err instanceof Error ? err.message : "Token refresh failed" });
      return false;
    }
  },

  // ── Webhooks ──
  registerWebhook: async (agentId: string, config: Record<string, unknown>) => {
    try {
      set({ error: null });
      const data = await api.post<WebhookRegistration>(
        `${API_PREFIX}/webhooks/register`,
        { agent_id: agentId, config }
      );
      await get().fetchWebhooks();
      return data;
    } catch (err) {
      set({ error: err instanceof Error ? err.message : "Failed to register webhook" });
      return null;
    }
  },

  unregisterWebhook: async (webhookId: string) => {
    try {
      set({ error: null });
      await api.post(`${API_PREFIX}/webhooks/${webhookId}/unregister`);
      await get().fetchWebhooks();
      return true;
    } catch (err) {
      set({ error: err instanceof Error ? err.message : "Failed to unregister webhook" });
      return false;
    }
  },

  fetchWebhooks: async (agentId?: string) => {
    try {
      const params = agentId ? { agent_id: agentId } : undefined;
      const data = await api.get<WebhookRegistration[]>(
        `${API_PREFIX}/webhooks`,
        { params }
      );
      set({ webhooks: data });
    } catch (err) {
      set({ error: err instanceof Error ? err.message : "Failed to fetch webhooks" });
    }
  },

  // ── Reset ──
  reset: () => {
    set({
      connectors: [],
      selectedIntegration: null,
      statusSummary: null,
      isConnecting: false,
      isConnected: false,
      error: null,
      webhooks: [],
    });
  },
}));

File: services/backend/app/api/v1/integrations.py
"""
Integration & Webhook API Endpoints — All /api/v1/integrations routes.

From spec Section 8.2 (Integrations):
- GET  /integrations — List all available connections with status
- GET  /integrations/:id — Integration detail
- POST /integrations/:id/auth — Start OAuth flow
- POST /integrations/:id/disconnect — Remove connection
- GET  /integrations/status — Health of all connections
- GET  /integrations/connectors — List available connector types

Additional:
- POST /integrations/:id/test — Test a connection
- POST /integrations/:id/refresh — Refresh OAuth token
- POST /integrations/connect/:key — Connect via API key

Webhook endpoints:
- POST /integrations/webhooks/register — Register webhook
- POST /integrations/webhooks/:id/unregister — Unregister webhook
- POST /webhooks/incoming/:id — Receive webhook (public, no auth)
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, Path, Query, Request
from pydantic import BaseModel, Field
from structlog import get_logger

from app.core.dependencies import get_current_user, CurrentUser
from app.services.integration import IntegrationService, get_integration_service
from app.services.webhook import WebhookService, get_webhook_service

logger = get_logger(__name__)
router = APIRouter()


# ─── Request / Response Models ──────────────────────────────────────────────


class OAuthInitiateRequest(BaseModel):
    connector_key: str = Field(..., description="Connector key (e.g. 'gmail', 'slack')")
    redirect_uri: str = Field(..., description="OAuth redirect URI for callback")


class OAuthCallbackRequest(BaseModel):
    connector_key: str = Field(..., description="Connector key")
    code: str = Field(..., description="Authorization code from OAuth provider")
    state: str = Field(..., description="State parameter for verification")


class ApiKeyConnectRequest(BaseModel):
    connector_key: str = Field(..., description="Connector key (e.g. 'notion', 'linear')")
    api_key: str = Field(..., description="API key or token for the service")
    config: dict[str, Any] | None = Field(default=None, description="Additional configuration")


class WebhookRegisterRequest(BaseModel):
    agent_id: str = Field(..., description="Agent UUID to trigger")
    config: dict[str, Any] = Field(default_factory=dict, description="Webhook configuration")


# ─── Integration Endpoints ──────────────────────────────────────────────────


@router.get("/integrations", summary="List all available connectors with connection status")
async def list_integrations(
    current_user: CurrentUser = Depends(get_current_user),
    svc: IntegrationService = Depends(get_integration_service),
):
    """List all available connector types with the user's connection status."""
    return await svc.list_available(current_user.id)


@router.get("/integrations/connectors", summary="List available connector types")
async def list_connectors(
    current_user: CurrentUser = Depends(get_current_user),
    svc: IntegrationService = Depends(get_integration_service),
):
    """List all available connector types."""
    available = await svc.list_available(current_user.id)
    # Return just the connector metadata without status
    return [
        {
            "key": c["key"],
            "name": c["name"],
            "description": c["description"],
            "icon_url": c["icon_url"],
            "category": c["category"],
            "auth_type": c["auth_type"],
            "scopes": c.get("scopes", []),
        }
        for c in available
    ]


@router.get("/integrations/status", summary="Health status of all connections")
async def get_integration_status(
    current_user: CurrentUser = Depends(get_current_user),
    svc: IntegrationService = Depends(get_integration_service),
):
    """Return the health of all user connections."""
    return await svc.get_status(current_user.id)


@router.get("/integrations/connectors/{key}", summary="Get connector type detail")
async def get_connector_detail(
    key: str = Path(..., description="Connector key"),
    current_user: CurrentUser = Depends(get_current_user),
    svc: IntegrationService = Depends(get_integration_service),
):
    """Get detailed info about a specific connector type."""
    return await svc.get_connector_detail(key)


@router.get("/integrations/{integration_id}", summary="Get integration detail")
async def get_integration_detail(
    integration_id: str = Path(..., description="Integration UUID"),
    current_user: CurrentUser = Depends(get_current_user),
    svc: IntegrationService = Depends(get_integration_service),
):
    """Get detailed info about a specific user integration."""
    return await svc.get_integration_detail(integration_id, current_user.id)


@router.post("/integrations/auth", summary="Start OAuth flow for a connector")
async def initiate_oauth(
    req: OAuthInitiateRequest,
    current_user: CurrentUser = Depends(get_current_user),
    svc: IntegrationService = Depends(get_integration_service),
):
    """Initiate OAuth flow. Returns the authorization URL to redirect the user to."""
    return await svc.initiate_oauth(
        connector_key=req.connector_key,
        user_id=current_user.id,
        redirect_uri=req.redirect_uri,
    )


@router.post("/integrations/auth/callback", summary="Complete OAuth flow")
async def oauth_callback(
    req: OAuthCallbackRequest,
    svc: IntegrationService = Depends(get_integration_service),
):
    """Complete the OAuth flow by exchanging the authorization code for tokens.

    This endpoint is called by the OAuth provider's redirect, or can be called
    from the frontend after the redirect is captured.
    """
    return await svc.handle_oauth_callback(
        connector_key=req.connector_key,
        code=req.code,
        state=req.state,
    )


@router.post("/integrations/connect", summary="Connect via API key")
async def connect_api_key(
    req: ApiKeyConnectRequest,
    current_user: CurrentUser = Depends(get_current_user),
    svc: IntegrationService = Depends(get_integration_service),
):
    """Connect a service using an API key (for services without OAuth)."""
    return await svc.connect_api_key(
        connector_key=req.connector_key,
        user_id=current_user.id,
        api_key=req.api_key,
        config=req.config,
    )


@router.post("/integrations/{integration_id}/disconnect", summary="Remove a connection")
async def disconnect_integration(
    integration_id: str = Path(..., description="Integration UUID"),
    current_user: CurrentUser = Depends(get_current_user),
    svc: IntegrationService = Depends(get_integration_service),
):
    """Disconnect and remove an integration, revoking tokens if possible."""
    return await svc.disconnect(integration_id, current_user.id)


@router.post("/integrations/{integration_id}/test", summary="Test a connection")
async def test_connection(
    integration_id: str = Path(..., description="Integration UUID"),
    current_user: CurrentUser = Depends(get_current_user),
    svc: IntegrationService = Depends(get_integration_service),
):
    """Test a connection by calling the service's endpoint."""
    return await svc.test_connection(integration_id, current_user.id)


@router.post("/integrations/{integration_id}/refresh", summary="Refresh OAuth token")
async def refresh_token(
    integration_id: str = Path(..., description="Integration UUID"),
    current_user: CurrentUser = Depends(get_current_user),
    svc: IntegrationService = Depends(get_integration_service),
):
    """Force refresh an OAuth token if it's expired or about to expire."""
    return await svc.refresh_token(integration_id, current_user.id)


# ─── Webhook Endpoints ─────────────────────────────────────────────────────


@router.post("/integrations/webhooks/register", summary="Register a webhook for an agent")
async def register_webhook(
    req: WebhookRegisterRequest,
    current_user: CurrentUser = Depends(get_current_user),
    svc: WebhookService = Depends(get_webhook_service),
):
    """Register a webhook endpoint for an agent. Returns the webhook URL and secret."""
    return await svc.register_webhook(
        agent_id=req.agent_id,
        webhook_config=req.config,
    )


@router.post("/integrations/webhooks/{webhook_id}/unregister", summary="Unregister a webhook")
async def unregister_webhook(
    webhook_id: str = Path(..., description="Webhook UUID"),
    current_user: CurrentUser = Depends(get_current_user),
    svc: WebhookService = Depends(get_webhook_service),
):
    """Remove a webhook registration."""
    return await svc.unregister_webhook(webhook_id)


@router.get("/integrations/webhooks", summary="List registered webhooks")
async def list_webhooks(
    agent_id: str | None = Query(default=None, description="Filter by agent"),
    current_user: CurrentUser = Depends(get_current_user),
    svc: WebhookService = Depends(get_webhook_service),
):
    """List all registered webhooks, optionally filtered by agent."""
    return await svc.list_webhooks(agent_id=agent_id)


# ─── Public Incoming Webhook Endpoint ──────────────────────────────────────


@router.post(
    "/webhooks/incoming/{webhook_id}",
    summary="Receive an incoming webhook (public)",
    include_in_schema=False,  # Hide from OpenAPI docs — this is a machine-to-machine endpoint
)
async def receive_webhook(
    webhook_id: str = Path(...),
    request: Request,
    svc: WebhookService = Depends(get_webhook_service),
):
    """Public endpoint for external services to send webhooks to.

    No authentication required. Signature verification is done via HMAC
    if a secret was configured during registration.
    """
    body = await request.body()
    headers = dict(request.headers)
    return await svc.handle_incoming(webhook_id, headers, body)
